fhir-search/src/main/java/com/ibm/fhir/search/parameters/IncludesUtil.java

```diff
@@ -0,0 +1,116 @@
+        }
+
+        return result;
+    }
+
+    /**
+     * @param compartmentMemberType
+     * @param searchHelper
+     * @return a map from resource target type to the list of reference search parameter codes that can target that type
+     * @throws Exception
+     */
+    static Map<ResourceType, List<String>> getSearchCodesByType(String compartmentMemberType, SearchHelper searchHelper) throws Exception {
+        Map<ResourceType, List<String>> codesByType = new HashMap<>();
+
+        Map<String, SearchParameter> allSearchParameters = searchHelper.getSearchParameters(compartmentMemberType);
+        for (Map.Entry<String, SearchParameter> entry : allSearchParameters.entrySet()) {
+            SearchParameter sp = entry.getValue();
+            if (sp.getType().getValueAsEnum() != Value.REFERENCE) {
+                continue;
+            }
+
+            String expression = sp.getExpression().getValue();
+            for (ResourceTypeCode target : sp.getTarget()) {
+                // The search parameter target types come from the possible types of the target reference
+                // and don't take into account any "is X" filters from within the expression,
+                // so we use this heuristic to avoid adding the code when expression filtering applies
+                // and does not include a filter clause for this particular target type.
+                if (expression.contains(".is(") || expression.contains(" is ")) {
+                    if (!expression.contains(".is(" + target.getValue()) && !expression.contains(" is " + target.getValue())) {
+                        continue;
+                    }
+                }
+
+                codesByType.computeIfAbsent(target.getValueAsEnum(), rt -> new ArrayList<>())
+                        .add(entry.getKey());
+            }
+        }
+        return codesByType;
+    }
+
+
+    /**
+     * @param compartmentMemberType the resource type that is the source of the includes
+     * @param subResourceType the target resource type of the includes; null
+     * @param allowedIncludes the list of include parameter values to allow
+     * @param codes the search parameter codes to include on
+     * @return A list of parameter values to use for the _include query parameter.
+     * @throws Exception
+     */
+    private static List<String> computeIncludes(String compartmentMemberType, ResourceType subResourceType,
+            List<String> allowedIncludes, List<String> codes) throws Exception {
+        List<String> includes = new ArrayList<>();
+
+        for (String code : codes) {
+            // Need to make sure the search parameter has not been excluded
+            String parameterName = compartmentMemberType + ":" + code + ":" + subResourceType.value();
+            String simplifiedParameterName = compartmentMemberType + ":" + code;
+
+            boolean isAllowed = allowedIncludes == null ||
+                    allowedIncludes.contains(parameterName) ||
```

Comment:
Calling List#contains within a loop. Might be more appropriate to use a Set<String> here.